File: C++/Shape.cpp
#include "Shape.h"

bool Shape::isClockwise() {
	long double sum = 0;
	for (int i = 0; i < vertices.size(); ++i) {
		auto& cur = vertices[i];
		auto& next = vertices[(i + 1) % vertices.size()];
		sum += ((next.get_x() - cur.get_x()) * (next.get_y() + cur.get_y()));
	}
	return sum > 0;
}

void Shape::checkShape() {
	if (!isClockwise())
		throw std::invalid_argument(std::format("{}: The sequence of vertices is not clockwise", name));
}

void Shape::move(_Vertex_t deltaX, _Vertex_t deltaY) {
	//for (Vertex& vtx: vertices)
	for (Vertex& vtx : *this)
		vtx += {deltaX, deltaY};
}

Vertex& Shape::operator[](size_t index) {
	return vertices.at(index);
}

std::ostream& operator<<(std::ostream& os, Shape& shp) {
	shp.showInfo(os);
	return os;
}

std::vector<Vertex>::iterator Shape::begin() {
	return vertices.begin();
}
std::vector<Vertex>::iterator Shape::end() {
	return vertices.end();
}

std::vector<Vertex>::const_iterator Shape::cbegin() {
	return vertices.cbegin();
}
std::vector<Vertex>::const_iterator Shape::cend() {
	return vertices.cend();
}

File: Coursework-WinForms/Shape.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace Coursework_WinForms {
	public abstract class Shape : ICloneable {
		// properties
		public string name { get; set; }
		public List<Vertex> vertices { get; protected set; }
		[JsonConverter(typeof(ColorConverter))] public Color color { get; set; }

		// indexer
		public Vertex this[int index] {
			get { return vertices.ElementAt(index); }
			protected set { vertices[index] = value; }
		}

		// ctors
		public Shape(string name, List<Vertex> vertices) {
			this.name = name;
			this.vertices = new List<Vertex>(vertices);
			color = Color.Black;
			checkShape();
		}

		public Shape(string name, params Vertex[] vertices) {
			this.name = name;
			color = Color.Black;

			if (vertices == null)
				throw new ArgumentNullException("Empty vertex list");
			this.vertices = new List<Vertex>(vertices);
			checkShape();
		}

		// methods
		bool isClockwise() {
			double sum = 0;
			for (int i = 0; i < vertices.Count; i++) {
				var cur = vertices[i];
				var next = vertices[(i + 1) % vertices.Count];
				sum += (next.x - cur.x) * (next.y + cur.y);
			}
			return sum > 0;
		}

		void checkShape() {
			if (!isClockwise())
				throw new ArgumentException($"{name}: The sequence of vertices is not clockwise");
		}

		virtual public void move(float deltaX, float deltaY) {
			for (int i = 0; i < vertices.Count; i++)
				vertices[i] += new Vertex(deltaX, deltaY);
		}

		// abstact methods
		public abstract object Clone();
		public abstract double square();
		public abstract JsonNode makeJson();

		// object methods
		public override int GetHashCode() {
			return name.GetHashCode();
		}

		public override string ToString() {
			string result = $"Name: {name}\n" +
			$"Type: {GetType().Name}\n" +
			$"Square: {square()}\n" +
			"Vertices:\n";
			foreach (Vertex vtx in vertices)
				result += "\t" + vtx.ToString() + "\n";

			return result;
		}
	}
}
